toolkit/tools/grapher/grapher.go

```diff
@@ -110,47 +110,80 @@ func addUnresolvedPackage(g *pkggraph.PkgGraph, pkgVer *pkgjson.PackageVer) (new
+// addNodesForPackage creates a "Run", "Build", and "Test" node for the package described
+// in the Package structure. Returns pointers to the build and run Nodes
 // created, or an error if one of the nodes could not be created.
-func addNodesForPackage(g *pkggraph.PkgGraph, pkgVer *pkgjson.PackageVer, pkg *pkgjson.Package) (newRunNode *pkggraph.PkgNode, newBuildNode *pkggraph.PkgNode, err error) {
-	nodes, err := g.FindExactPkgNodeFromPkg(pkgVer)
+func addNodesForPackage(g *pkggraph.PkgGraph, pkg *pkgjson.Package) (err error) {
+	var (
+		newRunNode   *pkggraph.PkgNode
+		newBuildNode *pkggraph.PkgNode
+		newTestNode  *pkggraph.PkgNode
+	)
+
+	nodes, err := g.FindExactPkgNodeFromPkg(pkg.Provides)
 	if err != nil {
 		return
 	}
+
 	if nodes != nil {
-		logger.Log.Warnf(`Duplicate package name for package %+v read from SRPM "%s" (Previous: %+v)`, pkgVer, pkg.SrpmPath, nodes.RunNode)
-		err = nil
+		logger.Log.Warnf(`Duplicate package name for package %+v read from SRPM "%s" (Previous: %+v)`, pkg.Provides, pkg.SrpmPath, nodes.RunNode)
 		if nodes.RunNode != nil {
 			newRunNode = nodes.RunNode
 		}
 		if nodes.BuildNode != nil {
 			newBuildNode = nodes.BuildNode
 		}
+		if nodes.TestNode != nil {
+			newTestNode = nodes.TestNode
+		}
 	}
 
 	if newRunNode == nil {
 		// Add "Run" node
-		newRunNode, err = g.AddPkgNode(pkgVer, pkggraph.StateMeta, pkggraph.TypeLocalRun, pkg.SrpmPath, pkg.RpmPath, pkg.SpecPath, pkg.SourceDir, pkg.Architecture, "<LOCAL>")
-		logger.Log.Debugf("Adding run node %s with id %d\n", newRunNode.FriendlyName(), newRunNode.ID())
+		newRunNode, err = g.AddPkgNode(pkg.Provides, pkggraph.StateMeta, pkggraph.TypeLocalRun, pkg.SrpmPath, pkg.RpmPath, pkg.SpecPath, pkg.SourceDir, pkg.Architecture, "<LOCAL>")
 		if err != nil {
 			return
 		}
+		logger.Log.Debugf("Adding run node %s with id %d\n", newRunNode.FriendlyName(), newRunNode.ID())
 	}
 
 	if newBuildNode == nil {
 		// Add "Build" node
-		newBuildNode, err = g.AddPkgNode(pkgVer, pkggraph.StateBuild, pkggraph.TypeLocalBuild, pkg.SrpmPath, pkg.RpmPath, pkg.SpecPath, pkg.SourceDir, pkg.Architecture, "<LOCAL>")
-		logger.Log.Debugf("Adding build node %s with id %d\n", newBuildNode.FriendlyName(), newBuildNode.ID())
+		newBuildNode, err = g.AddPkgNode(pkg.Provides, pkggraph.StateBuild, pkggraph.TypeLocalBuild, pkg.SrpmPath, pkg.RpmPath, pkg.SpecPath, pkg.SourceDir, pkg.Architecture, "<LOCAL>")
 		if err != nil {
 			return
 		}
+		logger.Log.Debugf("Adding build node %s with id %d\n", newBuildNode.FriendlyName(), newBuildNode.ID())
 	}
 
-	// A "run" node has an implicit dependency on its coresponding "build" node, encode that here.
+	// A "run" node has an implicit dependency on its corresponding "build" node, encode that here.
 	err = g.AddEdge(newRunNode, newBuildNode)
 	if err != nil {
-		logger.Log.Errorf("Adding edge failed for %+v", pkgVer)
+		logger.Log.Errorf("Adding run -> build edge failed for %+v", pkg.Provides)
+		return
+	}
+
+	if !pkg.RunTests {
```

Comment:
Do we want to always inject test nodes into the graph? I'm not sure either way... If we always inject them we can re-use the graph artifacts for a build with check on/off, but the graph files also get larger... and it makes the scheduler simpler if the nodes just don't exist.

Reply:
I think I'd continue with what we've been doing so far. I think the graph didn't contain any test-only information for `RUN_CHECK=n`. Or at least we didn't include the test-only BRs and that's the concept I relied on while adding this change. In other words: for regular builds, I'm currently favour a graph, which contains only the nodes relevant for the build we're running right now. Otherwise, I think things may get unnecessarily confusing when we debug build issues.